---
title: "Immunology project for computation biologists"
output: github_document
bibliography: library.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(dplyr)
library(tidyr)
library(tibble)
library(readr)
```

The goal of this project is to design a pooled CRISPR screen to study an immunological process of your choice.

## Introduction to pooled CRISPR screens combined with single-cell RNA sequencing

In a pooled CRISPR screen, cells are transfected by a library of vectors that all target different genes in different ways. Each of these vectors is barcoded so that we can later derive which cell was targeted by which vector. A couple of hours after transfection, cells are profiled using single-cell (RNA) sequencing techniques [@wagner_revealing_2016]. After some computational analyses, we derive for each cell it's expression profile and the vectors by which the cells were transfected.

Most often, a gene is knocked-out (loss-of-function), for example Perturb-Seq [@dixitPerturbseqDissectingMolecular2016] or CROP-Seq [@datlingerPooledCRISPRScreening2017]). But genes can also be targeted in other ways, for example by overexpression (gain-of-function) using CRISPRa [@wangMultiplexedActivationEndogenous2019]. The possibilities are basically endless; one might imagine that libraries will be designed to target individual point mutations [@anzaloneSearchandreplaceGenomeEditing2019].

Most screens until now were performed _in vitro_, but exploring _in vivo_ screens is probably not far off. The main limiting factor here is making sure that the correct cells are transfected; one solution for immunology might be to do an adoptive transfer of bone marrow cells which were transfected.

## The project

Doing a pooled screen can be very expensive. To get enough signal over noise, you need to profile enough cells (thousands) for each vector, so checking the effect of 100 vectors requires you to profie hundreds of thousands of cells, and with about \$0.5-\$1 per cell, .... you can do the calculation!

To lower the cost, we typically want to do a targeted screen, where we narrow down to a subset of genes, between 10 and 50. This may be based on function (e.g. transcription factors) and expression pattern (e.g. genes specific to alveolar macrophages).

In this example, we're going to design a screen for transcription factors that may be important for alveolar macrophages. For the project, you should design a screen for a different topic. Feel free to explore and be creative!

## Selecting potential genes based on function

```{r, message=FALSE}
library(org.Mm.eg.db)
```

GO:0003700 is the Gene Ontology identifier for "DNA-binding transcription factor activity"

You can explore the Gene Ontology at https://www.ebi.ac.uk/QuickGO/term/GO:0003700

Our expression dataset uses gene symbols, so we have to convert the entrez ids

```{r}
tfs_entrezid <- as.list(org.Mm.egGO2ALLEGS)$`GO:0003700`
tfs <- unique(AnnotationDbi::select(org.Mm.eg.db, keys=tfs_entrezid, columns=c("SYMBOL"), keytype="ENTREZID")$SYMBOL)
```

```{r}
length(tfs)
```


## Selecting potential genes based on expression

```{r, message=FALSE}
library(limma)
```

### Using ImmGen RNA-seq

The ImmGen consortium has generated bulk gene expression dataset of a lot of immune cells. These datasets are generated using a standardized protocol to limit the batch effects, and the exact FACS gating strategies are (in most cases) provided so that the sorting can be reproduced (e.g. http://rstats.immgen.org/Skyline/resources/Sorting_PDFs_OSMNP/Mo.6Chi11b+.APAP.36h.Lv.pdf).

The most extensive dataset was generated using microarrays:

- To browse the data: http://www.immgen.org/databrowser/index.html 
- Data can be downloaded at https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE15907 and https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE37448

Nowadays, these is also an extensive RNA-seq dataset available:

- Browsing: http://rstats.immgen.org/Skyline/skyline.html
- ImmGen ULI RNASeq, containing an overview of all immune populations: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE127267
- ImmGen MNP OpenSource, a community-based dataset because of the extensive heterogeneity and the expertise required for correct sorting [@OpensourceImmGenMononuclear2016]: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE122108

We download the MNP dataset here:
```{r}
# geo <- getGEO("GSE122108")
if (!file.exists("data/GSE122108_Gene_count_table.csv.gz")) {
  curl::curl_download("https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE122108&format=file&file=GSE122108_Gene_count_table.csv.gz", "data/GSE122108_Gene_count_table.csv.gz")
}
```

```{r}
# read in the counts matrix
counts_df <- read.table("data/GSE122108_Gene_count_table.csv.gz", header = TRUE, stringsAsFactors = FALSE)
counts <- as.matrix(counts_df[, 2:ncol(counts_df)])
rownames(counts) <- counts_df$gene_symbol
# colnames(counts) <- gsub('-', 'min', colnames(counts)) # R doesnt like - or + signs in names
# colnames(counts) <- gsub('\\+', 'plus', colnames(counts))

# remove genes that are very lowly expressed
counts <- counts[apply(counts, 1, max) >= 10, ]

# create a dataframe with information on each sample
sample_info <- data.frame(
  sample_id =  colnames(counts),
  population_id = gsub("(.*)\\.[0-9]*", "\\1", colnames(counts)),
  stringsAsFactors = FALSE
)
```

```{r}
head(sample_info, 20)
```

We can use limma-voom to find genes that are differentially expressed between two populations. Let's first explore how to get differentially expressed genes between two populations of interest, for example Kupffer cells and alveolar macrophages:

```{r}
population_ids_oi <- c("MF.alv.11cp64pSiglecFp.Lu","MF.KC.Clec4FpTim4p64p.Lv")
```

```{r, echo = FALSE}
sample_info_oi <- sample_info %>% 
  filter(population_id %in% population_ids_oi)
counts_oi <- counts[, sample_info_oi$sample_id]

group <- sample_info_oi$population_id
mm <- model.matrix(~0 + group)
y <- voom(counts_oi, mm, plot = F)

# construct a linear model with one term: the "group"

fit <- lmFit(y, mm)
contr <- makeContrasts(paste0("group", population_ids_oi[[1]], "-", "group", population_ids_oi[[2]]), levels = colnames(coef(fit)))
contr_fit <- contrasts.fit(fit, contr)
```

```{r, echo = FALSE}
ebayes_fit <- eBayes(contr_fit)

top.table <- topTable(ebayes_fit, sort.by = "P", number = Inf)
```

```{r, echo = FALSE}
genes_oi <- top.table %>% 
  rownames_to_column("gene_id") %>% 
  filter(adj.P.Val < 0.05, abs(logFC) > 1) %>% 
  top_n(100, abs(logFC)) %>% 
  pull(gene_id)
```

Plot the genes in a heatmap. We will scale the expression per gene.
```{r, echo = FALSE}
expression_oi <- edgeR::cpm(counts_oi)[genes_oi, ]
expression_oi_scaled <- t(scale(t(expression_oi)))
pheatmap::pheatmap(expression_oi_scaled)
```

We can do a differential expression for many different populations. The genes of interest are then for example the genes that are upregulated in our population of interest compared to all other "reference" populations. In our case, the reference are all types of macrophages.

```{r, echo = FALSE}
get_marker_genes <- function(counts, sample_info, populations_reference, population_oi) {
  # select relevant sample_info and counts
  sample_info_oi <- sample_info %>% 
    filter(population_id %in% c(populations_reference, population_oi))
  counts_oi <- counts[, sample_info_oi$sample_id]
  counts_oi <- counts_oi[apply(counts_oi, 1, mean) >= 1, ]
  
  # build linear model
  group <- sample_info_oi$population_id
  mm <- model.matrix(~0 + group)
  y <- voom(counts_oi, mm)
  
  fit <- lmFit(y, mm)
  
  # get differential expression for each reference population
  differential_expression <- lapply(populations_reference, function(population_reference) {
    contr_formula <- paste0("group", population_oi, "-", "group", population_reference)
    contr <- makeContrasts(contrasts = contr_formula, levels = colnames(coef(fit)))
    contr_fit <- contrasts.fit(fit, contr)
    
    ebayes_fit <- eBayes(contr_fit)
    top.table <- topTable(ebayes_fit, sort.by = "P", number = Inf)
    top.table %>% 
      rownames_to_column("gene_id") %>% 
      mutate(
        population_1 = population_oi,
        population_2 = population_reference
      )
  }) %>% 
    bind_rows()
}
```


```{r}
populations_reference <- c("MF.KC.Clec4FpTim4p64p.Lv", "MF.F.PC", "MF.480p.SP")
population_oi <- "MF.alv.11cp64pSiglecFp.Lu"
differential_expression <- get_marker_genes(counts, sample_info, populations_reference, population_oi)
```

```{r, echo = FALSE}
markers <- differential_expression %>% 
  group_by(gene_id) %>% 
  summarise(max_adj_pval = max(adj.P.Val), min_logfc = min(logFC)) %>% 
  filter(max_adj_pval < 0.05, min_logfc > 2) %>% 
  pull(gene_id)

samples_oi <- sample_info %>% filter(population_id %in% c(population_oi, populations_reference)) %>% pull(sample_id)
expression_oi <- log2(counts/colSums(counts) + 1)[markers, samples_oi]
expression_oi_scaled <- t(scale(t(expression_oi)))
pheatmap::pheatmap(expression_oi_scaled)
```

## Designing the CRISPR library

```{r}
genes_oi <- intersect(markers, tfs)
cat(paste0(genes_oi, collapse = "\n"))
```

We're lucky that we here get about 20 genes. What would you do if you get 500 genes? You might for example be a bit stricter and select the 20 genes that have the highest expression in our population of interest. What would you do if you get 2 genes? You can loosen the restrictions a bit and allow that the genes are also expressed in one or two of the reference populations.

We can submit this list to ... https://portals.broadinstitute.org/gpp/public/analysis-tools/sgrna-design (or any related tool). This will search for appropriate gRNA sequences. It will try to avoid off-targets by doing a BLAST search against the genome of interest, and selecting those gRNAs that are only found once, i.e. at the location of interest.

```{r}
sgrna_design <- read_tsv("results/sgrna_designs.txt")
```

Let's choose for each target gene one guide RNA:

```{r}
sgrna_design %>% 
  group_by(Input) %>% 
  top_n(1, `Combined Rank`) %>% 
  dplyr::select(`Target Gene Symbol`, `sgRNA Sequence`)
```

Time to start cloning!

## Analysis

To analyze the data, you need to know for each read the cell to which it belongs (cell barcode) and how this cell was perturbed (gRNA barcode).

Analyzing the data happens in two steps: preprocessing and modelling.

In the preprocessing step, you count the number of molecules that are part of each cell. You also assign to each cell with which vector it was transfected, so that you know which genes were affected in your cell.

In the modelling step, you have a look at the effect of the perturbation on your cell of interest. These analyses can be:

- Finding modules of genes that are affected by different perturbations. These modules can have functions, e.g. a module that is related to lipid metabolism.
- If you combine perturbations of different genes, you can also analyze interactions between genes, e.g. whether they work synergistically, antagonistically, ...

The current state-of-the-art of CRISPR screen modelling is probably: https://science.sciencemag.org/content/365/6455/786/tab-figures-data together.

## References